activerecord/lib/active_record/locking/pessimistic.rb

```diff
@@ -67,6 +67,10 @@ module Pessimistic
       # or pass true for "FOR UPDATE" (the default, an exclusive row lock). Returns
       # the locked record.
       def lock!(lock = true)
+        if self.class.current_preventing_writes
+          raise ActiveRecord::ReadOnlyError, "Lock query attempted while in readonly mode"
```

Comment:
This is the matching write query error for reference: https://github.com/rails/rails/blob/ffc8c47bee784ed635899765cdb8a52ff961b6f2/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L195